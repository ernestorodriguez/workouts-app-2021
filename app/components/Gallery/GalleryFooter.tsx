import React, { ReactElement } from "react";
import { Pagination } from "semantic-ui-react";

interface GalleryFooterProps {
  getPage(activePage: number): void;
  page?: number;
  totalPages?: number;
  totalWorkOuts?: number;
}

const GalleryFooter = ({
  getPage,
  page,
  totalPages,
  totalWorkOuts,
}: GalleryFooterProps): ReactElement => (
  <>
    <footer className="gallery-footer">
      <div data-js="total-workouts">{`${totalWorkOuts} Workouts`}</div>
      {totalPages && totalPages > 1 && (
        <Pagination
          className="pagination"
          activePage={page}
          firstItem={null}
          lastItem={null}
          boundaryRange={0}
          siblingRange={1}
          pointing
          secondary
          totalPages={totalPages}
          onPageChange={(event, data) => {
            getPage(data.activePage as number);
          }}
        />
      )}
    </footer>
  </>
);

export default GalleryFooter;
