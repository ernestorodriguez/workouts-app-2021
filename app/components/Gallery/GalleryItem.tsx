import React, {ReactElement, useState} from "react";
import { NavLink } from "react-router-dom";
import { Placeholder } from "semantic-ui-react";

export interface GalleryItemProps {
  id: number;
  name: string;
  alias: string;
  startDate: string;
  thumbnailMedium: string;
  thumbnailDefault: string;
  thumbnailHigh: string;
  category: string;
}

const GalleryItem = ({
  id,
  name,
  alias,
  startDate,
  thumbnailDefault,
  thumbnailMedium,
  thumbnailHigh,
  category,
}: GalleryItemProps): ReactElement => {
  const [loading, setLoading] = useState(true);
  return (
    <article className="gallery-item" data-js="gallery-item" data-id={id}>
      <NavLink className="gallery-item_link" exact to={`/workouts/${alias}`}>
        <div className="gallery-item__image-container">
          <img
            onLoad={() => {
              setLoading(false);
            }}
            srcSet={`${thumbnailMedium} 480w, ${thumbnailHigh} 800w`}
            sizes="(max-width: 600px) 480px, 800px"
            loading="lazy"
            className={`gallery-item__image ${
              loading && "gallery-item__image--loading"
            }`}
            src={thumbnailDefault}
            alt=""
          />
          {loading && (
            <Placeholder>
              <Placeholder.Image rectangular />
            </Placeholder>
          )}
        </div>

        <div className="gallery-item_info">
          <div className="gallery-item_info-name">{name}</div>
          <div data-js="start-date">{startDate}</div>
          <div className="gallery-item__category" data-js="start-category">
            {category}
          </div>
        </div>
      </NavLink>
    </article>
  );
}

export default GalleryItem;
