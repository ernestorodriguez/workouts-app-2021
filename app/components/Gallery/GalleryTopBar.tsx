import React, { ReactElement } from "react";
import { Dropdown, DropdownItemProps } from "semantic-ui-react";

interface TopBarProps {
  startDateChangeHandler(value: string): void;
  categoryChangeHandler(value: string[]): void;
  startDateSelector?: Record<string, unknown>;
  availableCategories?: string[];
}

const TopBar = ({
  startDateChangeHandler,
  categoryChangeHandler,
  startDateSelector = {},
  availableCategories = [],
}: TopBarProps): ReactElement => {
  const startDateOption: DropdownItemProps[] = Object.keys(
    startDateSelector
  ).map((key) => {
    const dropdownItem: DropdownItemProps = {
      key,
      text: key,
      value: startDateSelector[key] as string | number | boolean | undefined,
    };
    return dropdownItem;
  });

  const availableCategoriesOption = availableCategories.map((key) => ({
    key,
    text: key,
    value: key,
  }));

  return (
    <header className="gallery-header" data-js="top-bar">
      <Dropdown
        className="gallery-header__dropdown start-date-selector"
        placeholder="Select Month"
        fluid
        selection
        options={startDateOption}
        onChange={(event, data) => {
          startDateChangeHandler(data.value as string);
        }}
      />
      <Dropdown
        className="gallery-header__dropdown category-selector"
        placeholder="Select Categories"
        fluid
        multiple
        selection
        options={availableCategoriesOption}
        data-js="start-date-filter"
        onChange={(event, data) => {
          categoryChangeHandler(data.value as string[]);
        }}
      />
    </header>
  );
};

export default TopBar;
