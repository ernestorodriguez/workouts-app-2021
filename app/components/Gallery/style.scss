.gallery-item {
  position: relative;
  padding: 10px;
}

.gallery-item_link {
  text-decoration: none;
  position: relative;
  display: block;
}

.gallery-item__image-container {
  overflow: hidden;
  border-radius: 10px;
}

.gallery-item__image {
  width: 100%;
  height: auto;
  transform: scale(1.05);
}

.gallery-item__category {
  background-color: lighten($color: #ff7f66, $amount: 5);
  padding: 5px 10px;
  border-radius: 10px;
  display: inline-block;
  overflow: hidden;
  color: white;
  font-weight: 400;
  text-shadow: none;
  margin-top: 10px;
}


.gallery-item_info {
  position: absolute;
  bottom: 0;
  color: whitesmoke;
  margin: 20px;
  padding-left: 20px;
  border-left: 1px solid white;
  text-shadow: rgba(0, 0, 0, 0.6) 2px 1px 10px;
}

.gallery-item_info-name {
  font-size: 1.2em;
  font-weight: 700;
}

.gallery-footer {
  text-align: center;
  margin-bottom: 30px;
}

.gallery-header {
  padding: 10px;
  max-width: 450px;
  margin: 0 auto;
}

.gallery-header__dropdown {
  margin-bottom: 10px;
}

@media only screen and (min-width: 620px) {
  .gallery-item-container {
    display: flex;
    flex-wrap: wrap;
    padding: 5px;
  }

  .gallery-item {
    width: 50%;
    padding: 5px;
  }
}

@media only screen and (min-width: 690px) {
  .gallery-item-container {
    padding: 5px;
    flex-flow: row wrap;
    margin: 0 auto;
  }

  .gallery-item {
    width: 33.33%;
  }
}

@media only screen and (min-width: 1280px) {
  .gallery-item-container {
    max-width: 1920px;
  }

  .gallery-item {
    width: 25%;
  }
}
