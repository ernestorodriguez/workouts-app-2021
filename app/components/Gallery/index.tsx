import React, { ReactChild, ReactElement, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import GalleryItem, { GalleryItemProps } from "./GalleryItem";
import GalleryTopBar from "./GalleryTopBar";
import { getGalleryPage } from "../../client/actions";
import GalleryFooter from "./GalleryFooter";
import { RootState } from "../../client/reducers";

const Gallery = (): ReactElement => {
  const children: ReactChild[] = [];

  const {
    workouts = [],
    page,
    startDate,
    startDateSelector,
    availableCategories,
    selectedCategories,
    totalWorkOuts,
    totalPages,
  } = useSelector(({ gallery }: RootState) => gallery);

  const dispatch = useDispatch();

  useEffect(() => {
    if (workouts && workouts.length === 0) {
      dispatch(getGalleryPage(1, startDate, selectedCategories));
    }
  }, []);

  workouts.forEach(
    ({
      id,
      name,
      alias,
      startDate,
      thumbnailMedium,
      category,
    }: GalleryItemProps) => {
      children.push(
        <GalleryItem
          key={`gallery-item-${id}`}
          id={id}
          name={name}
          alias={alias}
          startDate={startDate}
          thumbnailMedium={thumbnailMedium}
          category={category}
        />
      );
    }
  );

  const getPageForPagination = (newPage: number) => {
    dispatch(getGalleryPage(newPage, startDate, selectedCategories));
  };

  const startDateChangeHandler = (newDate: string) => {
    dispatch(getGalleryPage(1, newDate, selectedCategories));
  };
  const categoryChangeHandler = (newSelectedCategories: string[]) => {
    dispatch(getGalleryPage(1, startDate, newSelectedCategories));
  };

  return (
    <div className="gallery">
      <GalleryTopBar
        startDateChangeHandler={startDateChangeHandler}
        categoryChangeHandler={categoryChangeHandler}
        startDateSelector={startDateSelector}
        availableCategories={availableCategories}
      />
      <section className="gallery-item-container" data-js="gallery">
        {children}
      </section>
      <GalleryFooter
        getPage={getPageForPagination}
        page={page}
        totalPages={totalPages}
        totalWorkOuts={totalWorkOuts}
      />
    </div>
  );
};

export default Gallery;
