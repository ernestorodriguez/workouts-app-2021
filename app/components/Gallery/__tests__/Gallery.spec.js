import React from 'react';
import { shallow } from 'enzyme';
import Gallery from '../index';

describe('Gallery component', () => {
  test.skip('should match snapshot', () => {
    const workouts = [];

    // eslint-disable-next-line no-plusplus
    for (let i = 0; i < 20; i++) {
      workouts.push({
        id: i,
        name: `workout-${i}`,
        alias: 'alias-path',
        startDate: '2020-10-24',
        thumbnailMedium: 'http://google.com',
        category: 'c1',
      });
    }
    const wrapper = shallow(
      <Gallery
        selectedCategories={[]}
        availableCategories={[]}
        startDateSelector={{}}
        startDate="2020-07"
        workouts={workouts}
        page={1}
        getPage={() => {}}
        totalWorkOuts={100}
        totalPages={5}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
