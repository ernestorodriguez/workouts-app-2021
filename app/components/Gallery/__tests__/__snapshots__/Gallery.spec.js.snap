// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery component should match snapshot 1`] = `
<div
  className="gallery"
>
  <TopBar
    avaibleCategories={Array []}
    categoryChangeHandler={[Function]}
    startDateChangeHandler={[Function]}
    startDateSelector={Object {}}
  />
  <section
    className="gallery-item-container"
    data-js="gallery"
  >
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={0}
      key="gallery-item-0"
      name="workout-0"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={1}
      key="gallery-item-1"
      name="workout-1"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={2}
      key="gallery-item-2"
      name="workout-2"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={3}
      key="gallery-item-3"
      name="workout-3"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={4}
      key="gallery-item-4"
      name="workout-4"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={5}
      key="gallery-item-5"
      name="workout-5"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={6}
      key="gallery-item-6"
      name="workout-6"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={7}
      key="gallery-item-7"
      name="workout-7"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={8}
      key="gallery-item-8"
      name="workout-8"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={9}
      key="gallery-item-9"
      name="workout-9"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={10}
      key="gallery-item-10"
      name="workout-10"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={11}
      key="gallery-item-11"
      name="workout-11"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={12}
      key="gallery-item-12"
      name="workout-12"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={13}
      key="gallery-item-13"
      name="workout-13"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={14}
      key="gallery-item-14"
      name="workout-14"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={15}
      key="gallery-item-15"
      name="workout-15"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={16}
      key="gallery-item-16"
      name="workout-16"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={17}
      key="gallery-item-17"
      name="workout-17"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={18}
      key="gallery-item-18"
      name="workout-18"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
    <GalleryItem
      alias="alias-path"
      category="c1"
      id={19}
      key="gallery-item-19"
      name="workout-19"
      startDate="2020-10-24"
      thumbnailMedium="http://google.com"
    />
  </section>
  <GalleryFooter
    getPage={[Function]}
    page={1}
    totalPages={5}
    totalWorkOuts={100}
  />
</div>
`;
